hypothesis-python/src/hypothesis/extra/array_api.py

```diff
@@ -265,32 +265,22 @@ def __init__(self, xp, elements_strategy, dtype, shape, fill, unique):
         self.unique = unique
         self.array_size = math.prod(shape)
         self.builtin = find_castable_builtin_for_dtype(xp, dtype)
-
-    def set_value(self, result, i, val, strategy=None):
-        strategy = strategy or self.elements_strategy
-        try:
-            result[i] = val
-        except TypeError as e:
-            raise InvalidArgument(
-                f"Could not add generated array element {val!r} "
-                f"of dtype {type(val)} to array of dtype {result.dtype}."
-            ) from e
-        self.check_set_value(val, result[i], strategy)
+        self.check_hist = set()
```

Comment:
This is pretty tricky!

- It's safe to share between multiple tests cases *if and only if* we never add disallowed elements to the set.  
  A comment to this effect would be very useful; it took me a while to convince myself it was OK.
- We have potentially unbounded memory usage, if e.g. you set `max_examples=10**9` and leave int64 arrays overnight.  This can be cheaply mitigated by checking once per array (at the start of `do_draw`) whether the set has more than say 100K elements, and clearing it if so.

Reply:
I've made a comment which hopefully clarifies how `check_hist` works, hope that's okay.

Thanks for catching the potential memory problems—I've made a reset mechanism at the start of `do_draw()` to that effect.
```python
if len(self.check_hist) >= 100_000:
    self.check_hist = set()
```
This does hurt the performance gains. I tried tracking the cache size in a separate attribute, but `len()` for sets is really quite fast that it's not worth the cognitive load, and also using a list + size attribute was comparable in performance. So right now this PR seems improve performance by 5%-10% for typical uses of `xps.arrays()`.

I wrote test cases to cover both points... they are both admittedly a bit funky heh.

Reply:
Hmm, I'd rather not have that drop.  Try a range up to say 10M?  I think 80MB overhead is probably acceptable if it's a nontrivial perf improvement for arrays!

If you're benchmarking it would also be interesting to see perf without any checks at all - in some cases we could manage up-front analysis proving that we don't need them.

Finally, sharing the `check_hist` between strategies would also improve the hit rate: perhaps a global `defaultdict(set)` keyed off the dtype, and assign to an instance variable in `__init__`?  (and then self.check_hist.clear()` instead of assigning a new set)

Reply:
> perhaps a global defaultdict(set) keyed off the dtype

Just noting there's a potential hashing problem here, although @asmeurer mentioned to me recently that it's probably safe to assume dtypes will be hashable.